### PR TITLE
feat: Add `public` semantic token modifier for public items

### LINKED ISSUE #8943: Add semantic token modifier for `pub` items
When writing code, [extra care is often needed](https://www.tedinski.com/2018/02/06/system-boundaries.html) when working on items that lie on crate boundaries. Therefore, it might be helpful to highlight these differently. Crucially, I think that `pub` items _from the crate of the file being highlighted_ should be given this modifier, but not from other crates; otherwise, all items from other crates would have this modifier – how would you access them if they weren’t `pub`?

### PR DESCRIPTION
**Pull request by arzg**
Closes #8943.

### CODE REVIEW

**Comment by Veykril** [comment]
Can you also add the modifier and a short description to the list here? https://github.com/rust-analyzer/rust-analyzer/blob/5b4589f4747f242c038f19f5f2b51b2e03f93252/crates/ide/src/syntax_highlighting.rs#L122-L143

**Comment by Veykril** [comment]
CI tests are complaining about non-linearity here(all three platforms) but I don't see where this change would be non-linear? 🤔

**Comment by arzg** [comment]
> Can you also add the modifier and a short description to the list here?

Done!

> CI tests are complaining about non-linearity here(all three platforms) but I don't see where this change would be non-linear? 🤔

Any ideas how to fix this? Would it be worth rerunning to make sure it wasn’t a fluke? Then again, it isn’t good to have tests that fail on a whim.

**Comment by lnicola** [comment]
Looks linear enough to me.

bors retry

**Comment by lnicola** [comment]
bors try

**Comment by lnicola** [comment]
bors try

**Comment by matklad** [comment]
Looks rather quadratic here:

plot


```diff
diff --git a/crates/ide/src/syntax_highlighting/tests.rs b/crates/ide/src/syntax_highlighting/tests.rs
index bb68181fd..4edda0501 100644
--- a/crates/ide/src/syntax_highlighting/tests.rs
+++ b/crates/ide/src/syntax_highlighting/tests.rs
@@ -327,32 +327,33 @@ fn benchmark_syntax_highlighting_long_struct() {
 
 #[test]
 fn syntax_highlighting_not_quadratic() {
-    if skip_slow_tests() {
-        return;
-    }
+    // if skip_slow_tests() {
+    //     return;
+    // }
 
     let mut al = AssertLinear::default();
-    while al.next_round() {
-        for i in 6..=10 {
-            let n = 1 << i;
+    // while al.next_round() {
+    for i in 1..=45 {
+        let n = i * 100;
 
-            let fixture = bench_fixture::big_struct_n(n);
-            let (analysis, file_id) = fixture::file(&fixture);
+        let fixture = bench_fixture::big_struct_n(n);
+        let (analysis, file_id) = fixture::file(&fixture);
 
-            let time = Instant::now();
+        let time = Instant::now();
 
-            let hash = analysis
-                .highlight(file_id)
-                .unwrap()
-                .iter()
-                .filter(|it| it.highlight.tag == HlTag::Symbol(SymbolKind::Struct))
-                .count();
-            assert!(hash > n as usize);
+        let hash = analysis
+            .highlight(file_id)
+            .unwrap()
+            .iter()
+            .filter(|it| it.highlight.tag == HlTag::Symbol(SymbolKind::Struct))
+            .count();
+        assert!(hash > n as usize);
+        println!("P! {:>4?} {:?}", n, time.elapsed().as_millis());
 
-            let elapsed = time.elapsed();
-            al.sample(n as f64, elapsed.as_millis() as f64);
-        }
+        let elapsed = time.elapsed();
+        // al.sample(n as f64, elapsed.as_millis() as f64);
     }
+    // }
 }
 
 #[test]

```

File: crates/ide/src/syntax_highlighting/tests.rs [hunk 1]
```diff
@@ -237,6 +237,14 @@ fn use_foo_items() {
     }
 }
 
+pub enum Bool { True,False }
```
Comment (lnicola) [inline crates/ide/src/syntax_highlighting/tests.rs:240]:
```suggestion
pub enum Bool { True, False }
```

**Comment by arzg** [comment]
Thanks for making the effort to create that graph! Would you have any idea what about this change makes highlighting quadratic?

**Comment by matklad** [comment]
Huh, looks like we are also quadratic on master...

**Comment by matklad** [comment]
Marking as broken window:

* we should fix the fact that we are quadratic
* we should fix our is quadratic detector

Not sure if I'll be able to look into this myself today though

**Comment by iDawer** [comment]
I suppose this one
https://github.com/rust-analyzer/rust-analyzer/blob/3cb3f1d17b54a7e144db495e073551049c9c482c/crates/hir_def/src/adt.rs#L227-L229
along with field traverse causes such behavior:
Used in `SourceAnalyzer::resolve_record_field` https://github.com/rust-analyzer/rust-analyzer/blob/3cb3f1d17b54a7e144db495e073551049c9c482c/crates/hir/src/source_analyzer.rs#L183-L188

Call hierarchy leads to `higlight::traverse` which is a huge loop over all the fields
https://github.com/rust-analyzer/rust-analyzer/blob/3cb3f1d17b54a7e144db495e073551049c9c482c/crates/ide/src/syntax_highlighting.rs#L306-L312

**Comment by matklad** [comment]
@arzg could you rebase? https://github.com/rust-analyzer/rust-analyzer/pull/9362 Should’ve fixed the issue!

**Comment by arzg** [comment]
Done!

**Comment by lnicola** [comment]
No luck, it seems.

**Comment by matklad** [comment]
Aha, it seems that this PR does introduce O(N^2) in the end. Running this in release:

```rust
#[test]
fn benchmark_syntax_highlighting_long_struct() {
    for i in 1..=20 {
        let n = i * 500;
        let fixture = bench_fixture::big_struct_n(n);
        let (analysis, file_id) = fixture::file(&fixture);
        let t = std::time::Instant::now();
        analysis.highlight(file_id).unwrap();
        eprintln!("{:3}: {:?}", i, t.elapsed());
    }
}
```

I get

```
  1: 21.157851ms
  2: 39.387636ms
  3: 57.04733ms
  4: 76.089806ms
  5: 94.733162ms
  6: 113.956429ms
  7: 131.232812ms
  8: 153.221424ms
  9: 173.878544ms
 10: 191.949289ms
 11: 215.087703ms
 12: 231.287055ms
 13: 251.732164ms
 14: 269.931029ms
 15: 293.694245ms
 16: 316.968239ms
 17: 335.799605ms
 18: 355.772114ms
 19: 389.92803ms
 20: 412.518353ms
```

with master and 

```
  1: 33.046146ms
  2: 90.745823ms
  3: 168.102535ms
  4: 273.206792ms
  5: 394.522778ms
  6: 552.906436ms
  7: 750.60107ms
  8: 925.746766ms
  9: 1.148798134s
 10: 1.431866023s
 11: 1.726566724s
 12: 2.07207502s
 13: 2.445697905s
 14: 3.007470561s
 15: 2.98714759s
 16: 3.565271835s
 17: 3.975792518s
 18: 4.369784765s
 19: 4.776471804s
 20: 5.336346217s
```

with this PR. Let me take a closer look here....

**Comment by matklad** [comment]
Yeah, it's pertty clear why this is quadratic: 

```rust
impl ItemScope {
    pub fn entries<'a>(&'a self) -> impl Iterator<Item = (&'a Name, PerNs)> + 'a { ... }
    pub fn visibility_of(&self, def: ModuleDefId) -> Option<Visibility> {
        self.name_of(ItemInNs::Types(def))
            .or_else(|| self.name_of(ItemInNs::Values(def)))
            .map(|(_, v)| v)
    }
    pub(crate) fn name_of(&self, item: ItemInNs) -> Option<(&Name, Visibility)> {
        for (name, per_ns) in self.entries() {
            if let Some(vis) = item.match_with(per_ns) {
                return Some((name, vis));
            }
        }
        None
    }
```

**Comment by matklad** [comment]
Fixed the problem in #9379. 

@arzg could you rebase this again? Also, thank you so much for uncovering so many performance issues :-)

**Comment by arzg** [comment]
Thank you for the thorough analysis. Finding performance issues wasn’t my intention, but I guess it worked out :)

**Comment by arzg** [comment]
bors try

**Comment by Veykril** [comment]
odd, the workflow isnt starting for some reason
bors try

**Pull request merged**